### PR TITLE
Use pipe to properly block clients when server hasn't entered event loop.

### PR DESCRIPTION
**Pull request by objmagic**
The previous way of using `usleep` cannot guarantee the clients are blocked when server hasn't entered event loop, which makes``heron/common/tests/cpp/network:http_unittest`` flaky.

This PR switches to using pipe. Server will send a byte to pipe prior to entering event loop. Clients will wait(block) until the byte is read on the other side of the pipe. This makes sure that clients are properly blocked.

Test done via script
```Bash
#!/bin/sh

for i in `seq 1 1000`;
do
    bazel test --config=darwin --no_cache_test_results heron/common/tests/cpp/network:http_unittest
    if [[ $? != 0 ]]; then
      exit 1
    fi
done
```

Thanks @congwang for suggestions.

### CODE REVIEW

File: heron/common/tests/cpp/network/http_unittest.cpp [hunk 1]
```diff
@@ -64,9 +65,24 @@ void TerminateServer(sp_uint32 port) {
 
 void StartTest(sp_uint32 nclients, sp_uint64 requests, sp_uint32 nkeys) {
   // start the server thread
-  std::thread sthread(start_http_server, SERVER_PORT, nkeys);
+  int fds[2];
 
-  ::usleep(1000);
+  // choose an arbitrary number for ``sent'' as token
+  int recv, sent = 19583;
+  if (pipe(fds) < 0) {
```
Comment (congwang) [inline heron/common/tests/cpp/network/http_unittest.cpp]: Use ::pipe().

File: heron/common/tests/cpp/network/http_unittest.cpp [hunk 2]
```diff
@@ -64,9 +65,24 @@ void TerminateServer(sp_uint32 port) {
...
+    std::cerr << "Unable to open pipe" << std::endl;
+    GTEST_FAIL();
+  }
+
+  std::thread sthread(start_http_server, SERVER_PORT, nkeys, fds[1], sent);
+
+  // block clients from reading from server using pipe
+  read(fds[0], &recv, sizeof(int));
+  if (recv != sent) {
```
Comment (congwang) [inline heron/common/tests/cpp/network/http_unittest.cpp]: No need to check the value, we only have one reader and one writer.

File: heron/common/tests/cpp/network/http_unittest.cpp (hunk 1 above)
Comment (objmagic) [inline heron/common/tests/cpp/network/http_unittest.cpp]: resolved via https://github.com/twitter/heron/pull/1541/commits/0238865309396146ecb897aac5eec9260b7015d8

File: heron/common/tests/cpp/network/http_unittest.cpp (hunk 2 above)
Comment (objmagic) [inline heron/common/tests/cpp/network/http_unittest.cpp]: resolved via https://github.com/twitter/heron/pull/1541/commits/593f35ad033a133938057f381844285804be373a

**Comment by kramasamy** [comment]
Good one! Thanks for fixing this.

File: heron/common/tests/cpp/network/http_unittest.cpp [hunk 3]
```diff
@@ -35,7 +36,7 @@ static const sp_uint32 SERVER_PORT = 60000;
 
 extern void start_http_client(sp_uint32 _port, sp_uint64 _requests, sp_uint32 _nkeys);
 
-extern void start_http_server(sp_uint32 _port, sp_uint32 _nkeys);
+extern void start_http_server(sp_uint32 _port, sp_uint32 _nkeys, int fd, int sent);
```
Comment (congwang) [inline heron/common/tests/cpp/network/http_unittest.cpp]: 'sent' is no longer needed either.

Comment (objmagic) [inline heron/common/tests/cpp/network/http_unittest.cpp]: resolved via https://github.com/twitter/heron/pull/1541/commits/dc185672ac8d8b9bd773ecf9246c4a32ef6a3bc0

**Comment by congwang** [comment]
Looks good to me.

**Comment by billonahill** [comment]
Should fix #1544.

**Pull request merged**